### PR TITLE
Add RealmSchema.createWithPrimaryKeyField()

### PR DESCRIPTION
**Pull request by beeender**
And deprecate RealmObject.addPrimaryKey()/removePrimaryKey().

### CODE REVIEW

**Comment by beeender** [comment]
This moves part of the `RealmSchema.createWithPrimaryKeyField()` implementation from #4693

File: realm/realm-library/src/main/java/io/realm/FieldAttribute.java [hunk 1]
```diff
@@ -33,7 +33,10 @@
      * Marks a field as a primary key. This also implicitly mark it as {@link #INDEXED} and {@link #REQUIRED}.
      *
      * @see io.realm.annotations.PrimaryKey
+     * @deprecated adding or removing primary key after object schema gets created will be disallowed in
+     * realm-java 4.0.0.
      */
+    @Deprecated
```
Comment (beeender) [inline realm/realm-library/src/main/java/io/realm/FieldAttribute.java]: I am not quite sure it is good to mark this as `@Deprecated`, currently the only public API using this is `addField()`, so removing this in the 4.0.0 is accepted? Or do we need this in the future API like `RealmObjectSchema.getFieldAttribute()`?

**Comment by Zhuinden** [comment]
So now it's only possible to make a `long` primary key field into a `String` primary key field is by renaming the class, creating a new class with all existing fields (and the new ID of course), using transform to map from the old object schema  to the new object schema,  and removing the old object schema? 

Harsh.

**Comment by beeender** [comment]
@Zhuinden yes, :( it will be very much like what you described above. however some helper function could be made to support that in a better way .

File: realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java [hunk 2]
```diff
@@ -115,6 +115,87 @@ public void create_duplicatedNameThrows() {
...
+        realmSchema.createWithPrimaryKeyField("FooNull", "pkField", Integer.class);
+        objectSchema = realmSchema.getSchemaForClass("FooNull");
+        assertEquals("pkField", objectSchema.getPrimaryKey());
+        assertEquals(RealmFieldType.INTEGER, objectSchema.getFieldType("pkField"));
+        assertTrue(objectSchema.isNullable("pkField"));
+    }
+
+    @Test
+    public void createWithPrimaryKeyField_invalidClassNameThrows() {
+        String[] names = { null, "", TestHelper.getRandomString(57) };
```
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java]: Shouldn't we avoid to have U+0000 or any other chars not valid for table name in the name?
I think we should update `TestHelper.getRandomString(int)` to exclude U+0000 at least.

Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java]: `invalidNames` is better than just `names`.

File: realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java [hunk 3]
```diff
@@ -115,6 +115,87 @@ public void create_duplicatedNameThrows() {
...
+                fail();
+            } catch (IllegalArgumentException ignored) {
+            }
+            assertFalse(String.format("'%s' failed", name), realmSchema.contains(name));
+        }
+    }
+
+    @Test
+    public void createWithPrimaryKeyField_invalidFieldNameThrows() {
+        String[] fieldNames = new String[] { null, "", "foo.bar", TestHelper.getRandomString(65) };
```
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java]: We must make sure that the return value of `TestHelper.getRandomString(65)` does not contain `.`.

File: realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java [hunk 4]
```diff
@@ -115,6 +115,87 @@ public void create_duplicatedNameThrows() {
...
+        assertEquals(RealmFieldType.INTEGER, objectSchema.getFieldType("pkField"));
+        assertFalse(objectSchema.isNullable("pkField"));
+
+        // Nullable
+        realmSchema.createWithPrimaryKeyField("FooNull", "pkField", Integer.class);
+        objectSchema = realmSchema.getSchemaForClass("FooNull");
+        assertEquals("pkField", objectSchema.getPrimaryKey());
+        assertEquals(RealmFieldType.INTEGER, objectSchema.getFieldType("pkField"));
+        assertTrue(objectSchema.isNullable("pkField"));
+    }
```
Comment (zaki50) [inline realm/realm-library/src/androidTest/java/io/realm/RealmSchemaTests.java]: We should have tests that pass `FieldAttribute.PRIMARY_KEY` or `FieldAttribute.INDEXED`.

File: realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java [hunk 5]
```diff
@@ -308,6 +308,24 @@ public Table createTable(String name) {
...
+     * @param isNullable if the primary key field is nullable or not.
+     * @return a creatd {@link Table} object.
+     */
+    public Table createTableWithPrimaryKey(String tableName, String primaryKeyFieldName, boolean isStringType,
+                                           boolean isNullable) {
+        return new Table(this, nativeCreateTableWithPrimaryKeyField(nativePtr, tableName, primaryKeyFieldName,
+                isStringType, isNullable));
+    }
+
+
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/internal/SharedRealm.java]: too many empty lines

File: realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java [hunk 6]
```diff
@@ -38,7 +38,7 @@
  */
 public class RealmObjectSchema {
 
-    private static final Map<Class<?>, FieldMetaData> SUPPORTED_SIMPLE_FIELDS;
+    static final Map<Class<?>, FieldMetaData> SUPPORTED_SIMPLE_FIELDS;
```
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java]: Consider a static method to access this.  Just exposing the implementation is a pretty drastic thing to do.

File: realm/realm-library/src/main/java/io/realm/RealmSchema.java [hunk 7]
```diff
@@ -27,6 +27,8 @@
 import io.realm.internal.Util;
 import io.realm.internal.util.Pair;
 
+import static io.realm.RealmObjectSchema.SUPPORTED_SIMPLE_FIELDS;
```
Comment (bmeike) [inline realm/realm-library/src/main/java/io/realm/RealmSchema.java]: Please consider a normal import, and qualification at each use.  It makes the code so much clearer...

File: realm/realm-library/src/main/java/io/realm/RealmSchema.java [hunk 8]
```diff
@@ -117,6 +116,42 @@ public RealmObjectSchema create(String className) {
...
+    public RealmObjectSchema createWithPrimaryKeyField(String className, String primaryKeyFieldName, Class<?> fieldType,
+                                                       FieldAttribute... attributes) {
+        checkEmpty(className, EMPTY_STRING_MSG);
+        RealmObjectSchema.checkLegalName(primaryKeyFieldName);
+        String internalTableName = checkAndGetTableNameFromClassName(className);
+
+        RealmObjectSchema.FieldMetaData metadata = RealmObjectSchema.getSupportedSimpleFields().get(fieldType);
+        if (metadata == null || (metadata.realmType != RealmFieldType.STRING &&
+                metadata.realmType != RealmFieldType.INTEGER)) {
+            throw new IllegalArgumentException(String.format("Realm doesn't support primary key field type '%s'.",
```
Comment (zaki50) [inline realm/realm-library/src/main/java/io/realm/RealmSchema.java]: Locale (`Locale.US`) is missing.

File: realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp [hunk 9]
```diff
@@ -455,6 +457,38 @@ JNIEXPORT jlong JNICALL Java_io_realm_internal_SharedRealm_nativeCreateTable(JNI
...
+    try {
+        std::string table_name(JStringAccessor(env, j_table_name));
+        class_name_str = std::string(table_name.substr(TABLE_PREFIX.length()));
+        JStringAccessor field_name(env, j_field_name); // throws
+        auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+        shared_realm->verify_in_write(); // throws
+        DataType pkType = is_string_type ? DataType::type_String : DataType::type_Int;
+        Table* table;
+        auto& group = shared_realm->read_group();
+        table = LangBindHelper::add_table(group, table_name);
```
Comment (kneth) [inline realm/realm-library/src/main/cpp/io_realm_internal_SharedRealm.cpp]: Doesn't it make sense to have this and the next lines as a function in object store? I mean, to ensure we create primary key fields the same way?

**Comment by beeender** [comment]
Deprecating those may create troubles for current users. close this for now. Maybe we can try to avoid those for non-sync users.

**Pull request closed without merging**